### ISSUE TITLE
Hotkey support

### ISSUE THREAD

**Issue by cedrics** [allegro/turnilo#448]
We had hotkey support in our internal pivot / swiv fork and ported these changes over to turnilo when switching. We implemented hotkey support in a hacky way that you surely wouldn't have been merged 😄. However, our hacky implementation requires quiet some work to merge when you release new features, so we are thinking about building a real feature out of it and wanted to know if you have thought about hot key support before.

The hotkeys we added are:

- List navigation via. arrow keys and list item selection (for straight forward list via enter (for example. for filters) or further hotkeys for dimension selection (`s` for adding a split and `f` for adding a filter).
- Global hotkeys to focus the dimension and measure selection (open the search box & focus it)

Is this sth. you were interested in and if so what would you expect in terms of configuration?

Thanks for all the work you are doing!

**Comment by adrianmroz** [comment]
Hi!

Idea seems great! We need to discuss what keys will do what and where. For example if I click "+" on filters row, it would be nice to have arrows linked to up/down movement and enter to confirming selection. So we are open. 

Of course few reservations:
* hotkeys should always be alternative way to accomplish things, mouse support is first
* global hotkeys should be behind flag (probably configuration, maybe something in local storage?) because we're in browser and keyboard support there is ... a lot to be desired :)
* be careful of conflicts with focused inputs

List navigation for sure needs some UX thought - how to highlight selection, how to handle nested lists etc. 
Global hotkeys should be easy to add for developer. I'm thinking about component that ads listeners (like `GlobalEventListener`) with some API contract on `Essence`. 

Anyways, let's talk and discuss!

**Comment by mkuthan** [comment]
It would be a wonderful contribution. To limit the scope of the change we could think about a few milestones:
1. arrow keys support for dimension/measures panels
2. enter key for all dialogs with default action (e.g OK)
3. tab keys support for navigation between header menu, left panels, filters, splits, measures, etc. (in general from top-left to bottom-right)
4. tab keys support for dialogs (filter, split, measure), again from top-left to bottom-right
5. special shortcuts for adding filter, split or measure - for me the key bindings could be fixed but we need to prepare cheat sheet for Turnilo users and ensure that they do not collide with typical OS shortcuts

**Comment by cedrics** [comment]
Hey, that sounds great! What do you think about building keyboard support for a dialog with list selection as a first milestone? Main example would be adding a new filter. 

We currently use the arrow keys to mark an item in the list and then use space to select it (which is not ideal as you can't type a space in the search bar anymore). I would propose using enter to select an item and shift+enter to confirm the selection / close the dialog. WDYT? Slack uses an API where enter adds an item if there is one selected but commits the whole thing if none is selected (which could also be an alternative).

**Comment by mkuthan** [comment]
My milestones are only the proposal just to avoid many complex changes at once. Feel free to plan your work as you wish as long as there is no strict dependencies between milestones.

Let's talk about filter dialog. When it is opened the search field is focused, then you could use "tab" to navigate to the list of dimension values, then use "arrows" to navigate on the list and "space" for item selection/unselection. "Enter"/"Esc" bindings could be left as is - global action for the filter dialog.

For me "shift-enter" is well suited for dialogs with text area when "enter" binding is reserved for new line  in edited text.

**Comment by adrianmroz** [comment]
Just to be sure, which "filter dialog" are you talking about? `SelectableStringFilterMenu`? 
![image](https://user-images.githubusercontent.com/14370131/59669616-d67d6f00-91ba-11e9-9139-801e6b28590c.png)

**Comment by cedrics** [comment]
Hey, yes that is the menu I mean. 

> When it is opened the search field is focused, then you could use "tab" to navigate to the list of dimension values, then use "arrows" to navigate on the list and "space" for item selection/unselection. "Enter"/"Esc" bindings could be left as is - global action for the filter dialog.

So you are proposing that we have two different modes, either you type and filter the list or you are selecting items? This would make it a bit easier to implement but a bit more complex to use. Could a good starting point. My main concern would be how to jump back to the text box. Using tab to jump back would work nicely for now but may lead to problems if we have multiple focusable elements on the page in the future. Jumping between text box and selection quickly (e.g. filter done -> select -> repeat) could become a chore if you have to jump over multiple other focusable elements before you get back to the filter input.

**Comment by adrianmroz** [comment]
No, we shouldn't hijack `Tab` in browser. It is recipe for disaster or at least a pleathora of edge cases.
